api/metadata_handler.go

```diff
@@ -0,0 +1,184 @@
+type MetadataSearchHandler struct {
+	logger      shared.Logger
+	httpClient  *http.Client
+	metadataURL string
+}
+
+// apiMetadataHandler searches Metadata for given products.
+func apiMetadataHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" && r.Method != "POST" {
+		http.Error(w, "Invalid HTTP method", http.StatusBadRequest)
+		return
+	}
+
+	ctx := shared.NewAppEngineContext(r)
+	client := shared.NewAppEngineAPI(ctx).GetHTTPClient()
+	logger := shared.GetLogger(ctx)
+	metadataURL := shared.MetadataArchiveURL
+
+	var delegate http.Handler
+	if r.Method == "GET" {
+		delegate = MetadataHandler{logger, client, metadataURL}
+	} else {
+		delegate = MetadataSearchHandler{logger, client, metadataURL}
+	}
+
+	// Serve cached with 5 minute expiry. Delegate to Metadata Handler on cache miss.
+	shared.NewCachingHandler(
+		ctx,
+		delegate,
+		shared.NewGZReadWritable(shared.NewMemcacheReadWritable(ctx, 5*time.Minute)),
+		shared.AlwaysCachable,
+		cacheKey,
+		shared.CacheStatusOK).ServeHTTP(w, r)
+}
+
+func (h MetadataHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	q := r.URL.Query()
+	var productSpecs shared.ProductSpecs
+	productSpecs, err := shared.ParseProductOrBrowserParams(q)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	} else if len(productSpecs) == 0 {
+		http.Error(w, fmt.Sprintf("Missing required 'product' param"), http.StatusBadRequest)
+		return
+	}
+
+	metadataResponse, err := shared.GetMetadataResponseOnProducts(productSpecs, h.httpClient, h.logger, h.metadataURL)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	marshalled, err := json.Marshal(metadataResponse)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	_, err = w.Write(marshalled)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (h MetadataSearchHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "Failed to read request body", http.StatusInternalServerError)
+	}
+	err = r.Body.Close()
+	if err != nil {
+		http.Error(w, "Failed to finish reading request body", http.StatusInternalServerError)
+	}
+
+	var rq query.RunQuery
+	err = json.Unmarshal(data, &rq)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	var abstractLink query.AbstractLink
+	var isLinkQuery = false
+	if exists, isExists := rq.AbstractQuery.(query.AbstractExists); isExists && len(exists.Args) == 1 {
+		abstractLink, isLinkQuery = exists.Args[0].(query.AbstractLink)
+	}
+
+	if !isLinkQuery {
+		h.logger.Errorf("Error from request: non Link search query %s for api/metadata", rq.AbstractQuery)
+		http.Error(w, "Error from request: non Link search query for api/metadata", http.StatusInternalServerError)
```

Comment:
```suggestion
		http.Error(w, "Error from request: non Link search query for api/metadata", http.StatusBadRequest)
```
This is a user error.